src/ore/src/tracing.rs

```diff
@@ -145,58 +183,95 @@ where
+                        Resource::new([KeyValue::new("service.name", service_name.to_string())])
+                            .merge(&otel_config.resource),
+                    ),
+                )
+                .with_exporter(exporter)
+                .install_batch(opentelemetry::runtime::Tokio)
+                .unwrap();
+            let (otel_filter, filter_reload_handle) = reload::Layer::new(if otel_config.start_enabled {
+                otel_config.filter
+            } else {
+                // the default `Targets` has everything disabled
+                Targets::default()
             });
-        let exporter = opentelemetry_otlp::new_exporter()
-            .tonic()
-            .with_channel(channel)
-            .with_metadata(MetadataMap::from_headers(otel_config.headers));
-        let tracer = opentelemetry_otlp::new_pipeline()
-            .tracing()
-            .with_trace_config(
-                trace::config().with_resource(
-                    // The latter resources wins, so if the user specifies `service.name` on the
-                    // cli, it wins
-                    Resource::new([KeyValue::new("service.name", service_name.to_string())])
-                        .merge(&otel_config.resource),
-                ),
-            )
-            .with_exporter(exporter)
-            .install_batch(opentelemetry::runtime::Tokio)
-            .unwrap();
-        let layer = tracing_opentelemetry::layer()
-            .with_tracer(tracer)
-            .with_filter(otel_config.filter);
-        Some(layer)
-    } else {
-        None
+            let layer = tracing_opentelemetry::layer()
+                .with_tracer(tracer)
+                .with_filter(otel_filter);
+            Ok((Some(layer), Some(filter_reload_handle)))
+        } else {
+            Ok((None, None))
+        }
     };
 
+    let (otel_layer, reload_handle) = build_otel_layer(config.opentelemetry.clone())?;
+
+    let reloader = OpenTelemetryEnableCallback(Arc::new(move |enable| {
+        match (&reload_handle, &config.opentelemetry) {
+            (_, None) => Err(anyhow::anyhow!(
+                "Tried to {} the otel collector but there is no endpoint",
+                if enable { "enable" } else { "disable" }
+            )),
+            (Some(reload_handle), Some(otel_config)) => {
+                reload_handle.modify(|filter| {
+                    // This code should be kept panic-free to
+                    // avoid weird poisoning issues in our subscriber
+                    // stack.
+                    if enable {
+                        *filter = otel_config.filter.clone()
+                    } else {
+                        *filter = Targets::default()
+                    }
+                })?;
+                Ok(())
+            }
+            (None, Some(_)) => {
+                // TODO: structure this more clearly
+                unreachable!()
+            }
+        }
+    }));
```

Comment:
This looks like it comes out more cleanly if you instead create two different reload callbacks:

```diff
diff --git a/src/ore/src/tracing.rs b/src/ore/src/tracing.rs
index ecb3177ba..9e6bcb290 100644
--- a/src/ore/src/tracing.rs
+++ b/src/ore/src/tracing.rs
@@ -27,6 +27,7 @@ use std::net::SocketAddr;
 use std::sync::Arc;
 use std::time::Duration;
 
+use anyhow::bail;
 #[cfg(feature = "tokio-console")]
 use console_subscriber::ConsoleLayer;
 use http::HeaderMap;
@@ -183,94 +184,82 @@ where
         .with_ansi(atty::is(atty::Stream::Stderr))
         .with_filter(config.stderr_log.filter);
 
-    let build_otel_layer = move |otel_config: Option<OpenTelemetryConfig>|
-          -> Result<(Option<_>, Option<_>), anyhow::Error> {
-        if let Some(otel_config) = otel_config {
-            // TODO(guswynn): figure out where/how to call
-            // opentelemetry::global::shutdown_tracer_provider();
-            opentelemetry::global::set_text_map_propagator(TraceContextPropagator::new());
+    let (otel_layer, otel_reloader) = if let Some(otel_config) = config.opentelemetry {
+        // TODO(guswynn): figure out where/how to call
+        // opentelemetry::global::shutdown_tracer_provider();
+        opentelemetry::global::set_text_map_propagator(TraceContextPropagator::new());
 
-            // Manually set up an OpenSSL-backed, h2, proxied `Channel`,
-            // with the timeout configured according to:
-            // https://docs.rs/opentelemetry-otlp/latest/opentelemetry_otlp/struct.TonicExporterBuilder.html#method.with_channel
-            let channel = Endpoint::from_shared(otel_config.endpoint)?
-                .timeout(Duration::from_secs(
-                    opentelemetry_otlp::OTEL_EXPORTER_OTLP_TIMEOUT_DEFAULT,
-                ))
-                // TODO(guswynn): investigate if this should be non-lazy.
-                .connect_with_connector_lazy({
-                    let mut http = HttpConnector::new();
-                    http.enforce_http(false);
-                    HttpsConnector::from((
-                        http,
-                        // This is the same as the default, plus an h2 ALPN request.
-                        tokio_native_tls::TlsConnector::from(
-                            native_tls::TlsConnector::builder()
-                                .request_alpns(&["h2"])
-                                .build()
-                                .unwrap(),
-                        ),
-                    ))
-                });
-            let exporter = opentelemetry_otlp::new_exporter()
-                .tonic()
-                .with_channel(channel)
-                .with_metadata(MetadataMap::from_headers(otel_config.headers));
-            let tracer = opentelemetry_otlp::new_pipeline()
-                .tracing()
-                .with_trace_config(
-                    trace::config().with_resource(
-                        // The latter resources wins, so if the user specifies `service.name` on the
-                        // cli, it wins
-                        Resource::new([KeyValue::new("service.name", service_name.to_string())])
-                            .merge(&otel_config.resource),
+        // Manually set up an OpenSSL-backed, h2, proxied `Channel`,
+        // with the timeout configured according to:
+        // https://docs.rs/opentelemetry-otlp/latest/opentelemetry_otlp/struct.TonicExporterBuilder.html#method.with_channel
+        let channel = Endpoint::from_shared(otel_config.endpoint)?
+            .timeout(Duration::from_secs(
+                opentelemetry_otlp::OTEL_EXPORTER_OTLP_TIMEOUT_DEFAULT,
+            ))
+            // TODO(guswynn): investigate if this should be non-lazy.
+            .connect_with_connector_lazy({
+                let mut http = HttpConnector::new();
+                http.enforce_http(false);
+                HttpsConnector::from((
+                    http,
+                    // This is the same as the default, plus an h2 ALPN request.
+                    tokio_native_tls::TlsConnector::from(
+                        native_tls::TlsConnector::builder()
+                            .request_alpns(&["h2"])
+                            .build()
+                            .unwrap(),
                     ),
-                )
-                .with_exporter(exporter)
-                .install_batch(opentelemetry::runtime::Tokio)
-                .unwrap();
-            let (otel_filter, filter_reload_handle) = reload::Layer::new(if otel_config.start_enabled {
-                otel_config.filter
-            } else {
-                // the default `Targets` has everything disabled
-                Targets::default()
+                ))
             });
-            let layer = tracing_opentelemetry::layer()
-                .with_tracer(tracer)
-                .with_filter(otel_filter);
-            Ok((Some(layer), Some(filter_reload_handle)))
+        let exporter = opentelemetry_otlp::new_exporter()
+            .tonic()
+            .with_channel(channel)
+            .with_metadata(MetadataMap::from_headers(otel_config.headers));
+        let tracer = opentelemetry_otlp::new_pipeline()
+            .tracing()
+            .with_trace_config(
+                trace::config().with_resource(
+                    // The latter resources wins, so if the user specifies `service.name` on the
+                    // cli, it wins
+                    Resource::new([KeyValue::new("service.name", service_name.to_string())])
+                        .merge(&otel_config.resource),
+                ),
+            )
+            .with_exporter(exporter)
+            .install_batch(opentelemetry::runtime::Tokio)
+            .unwrap();
+        let (filter, filter_handle) = reload::Layer::new(if otel_config.start_enabled {
+            otel_config.filter.clone()
         } else {
-            Ok((None, None))
-        }
-    };
-
-    let (otel_layer, reload_handle) = build_otel_layer(config.opentelemetry.clone())?;
-
-    let reloader = OpenTelemetryEnableCallback(Arc::new(move |enable| {
-        match (&reload_handle, &config.opentelemetry) {
-            (_, None) => Err(anyhow::anyhow!(
+            // the default `Targets` has everything disabled
+            Targets::default()
+        });
+        let layer = tracing_opentelemetry::layer()
+            .with_tracer(tracer)
+            .with_filter(filter);
+        let reloader = OpenTelemetryEnableCallback(Arc::new(move |enable| {
+            filter_handle.modify(|filter| {
+                // This code should be kept panic-free to
+                // avoid weird poisoning issues in our subscriber
+                // stack.
+                if enable {
+                    *filter = otel_config.filter.clone()
+                } else {
+                    *filter = Targets::default()
+                }
+            })?;
+            Ok(())
+        }));
+        (Some(layer), reloader)
+    } else {
+        let reloader = OpenTelemetryEnableCallback(Arc::new(move |enable| {
+            bail!(
                 "Tried to {} the otel collector but there is no endpoint",
-                if enable { "enable" } else { "disable" }
-            )),
-            (Some(reload_handle), Some(otel_config)) => {
-                reload_handle.modify(|filter| {
-                    // This code should be kept panic-free to
-                    // avoid weird poisoning issues in our subscriber
-                    // stack.
-                    if enable {
-                        *filter = otel_config.filter.clone()
-                    } else {
-                        *filter = Targets::default()
-                    }
-                })?;
-                Ok(())
-            }
-            (None, Some(_)) => {
-                // TODO: structure this more clearly
-                unreachable!()
-            }
-        }
-    }));
+                if enable { "enable" } else { "disable" },
+            );
+        }));
+        (None, reloader)
+    };
 
     #[cfg(feature = "tokio-console")]
     let tokio_console_layer = if let Some(console_config) = config.tokio_console.clone() {
@@ -299,7 +288,7 @@ where
         );
     }
 
-    Ok(reloader)
+    Ok(otel_reloader)
 }
 
 /// Shutdown any tracing infra, if any.
```